src/nvtxw.rs

```diff
@@ -0,0 +1,427 @@
+                */
+            }
+        }
+    }
+    walk(
+        info,
+        EntryID::root(),
+        &mut result,
+        LEGION_DOMAIN_NAME.to_string(),
+    );
+    result
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct legion_nvtxw_event {
+    time_start: u64,
+    time_stop: u64,
+    name: *const c_char,
+    color: u32,
+}
+
+// See nvToolsExtPayload.h: nvtxPayloadSchemaAttr_t::schemaId
+// See NVTX_PAYLOAD_ENTRY_TYPE_SCHEMA_ID_STATIC_START
+const LEGION_NVTXW_PAYLOAD_SCHEMA_ID: u64 = 0x1c0ffee;
+const LEGION_NVTXW_PAYLOAD_NAME_SCHEMA_ID: u64 = 0x2c0ffee;
+
+impl<T: DeferredDataSource> NVTXW<T> {
+    pub fn new(
+        data_source: T,
+        output: Option<PathBuf>,
+        merge: Option<PathBuf>,
+        zero_time: i64,
+    ) -> Self {
+        Self {
+            data_source: CountingDeferredDataSource::new(data_source),
+            output,
+            merge,
+            zero_time,
+        }
+    }
+
+    fn check_info(&mut self) -> Option<DataSourceInfo> {
+        // We requested this once, so we know we'll get zero or one result
+        self.data_source.get_infos().pop()
+    }
+
+    fn write_matched_tile(
+        interface: nvtxw::InterfaceHandle,
+        streams: &BTreeMap<EntryID, nvtxw::StreamHandle>,
+        zero_time: i64,
+        tile: &SlotTile,
+        meta_tile: &SlotMetaTile,
+    ) {
+        assert!(tile.data.items.len() == meta_tile.data.items.len());
+
+        zip(&tile.data.items, &meta_tile.data.items).for_each(|(row, meta_row)| {
+            assert!(row.len() == meta_row.len());
+
+            zip(row, meta_row).for_each(|(item, meta_item)| {
+                let time_start = item.interval.start;
+                let time_stop = item.interval.stop;
+                let color = item.color;
+                // let time_start = meta_item.original_interval.start;
+                // let time_stop = meta_item.original_interval.stop;
+                let title = meta_item.title.clone();
+
+                let c_name = CString::new(title).expect("CString::new failed");
+                let events = [legion_nvtxw_event {
+                    time_start: (time_start.0 as u64)
+                        .checked_add(zero_time.try_into().unwrap())
+                        .expect("time_start overflowed"),
+                    time_stop: (time_stop.0 as u64)
+                        .checked_add(zero_time.try_into().unwrap())
+                        .expect("time_stop overflowed"),
+                    name: c_name.as_ptr(),
+                    color: ((color.r() as u32) << 16)
+                        | ((color.g() as u32) << 8)
+                        | (color.b() as u32)
+                        | (0xFF << 24),
+                }];
+
+                let stream = streams[&tile.entry_id];
+
+                let payloads = [
+                    nvtxw::PayloadData {
+                        schemaId: LEGION_NVTXW_PAYLOAD_NAME_SCHEMA_ID,
+                        size: usize::MAX,
```

Comment:
Is this supposed to be `size_of::<usize>()`?

Reply:
`usize::MAX` functions correctly here. In the C API, it would be `SIZE_MAX`, `(size_t)-1`, or `~(size_t)0`, etc. The NVTX extended payload spec uses this value as a sentinel to mean the backend should call `strlen` on the C string passed in.